### PR TITLE
Working meters

### PR DESCRIPTION
**Pull request by CharlieHardwick-Kelly**
Hope I'm doing this with the nmc/floodlight  and not the actual floodlight project, if this is the main floodlight please ignore

### CODE REVIEW

File: src/main/java/net/floodlightcontroller/core/web/serializers/NodeFlowTupleSerializer.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    @Override
+    public void serialize(NodeFlowTuple nft, JsonGenerator jGen,SerializerProvider serializer) throws IOException, JsonProcessingException {
+        serialize(nft, jGen);
+    }
+    
+    public static void serialize(NodeFlowTuple nft, JsonGenerator jGen) throws IOException {
+    
+    	
+    	//TO-DO need to update this with flow data
```
Comment (nmc16) [inline src/main/java/net/floodlightcontroller/core/web/serializers/NodeFlowTupleSerializer.java]: Is this an actual todo or just left there

File: src/main/java/net/floodlightcontroller/meters/Meters.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+	 */
+	public void addMeter(Match match, long meterID, IOFSwitch sw, List<OFInstruction> instr, TableId table) throws InterruptedException {
+		// need a list to hold all the built commands 
+		List<OFMeterBand> bands = new ArrayList<OFMeterBand>();
+		// First build the meter  
+		OFFactory meterFactory = OFFactories.getFactory(OFVersion.OF_13); 
+		OFMeterMod.Builder meterModBuilder = meterFactory.buildMeterMod().setMeterId(meterID).setCommand(OFMeterModCommand.ADD);		
+		OFMeterBand.Builder meterband = meterFactory.meterBands().buildExperimenter();
+		OFMeterBand band = meterband.build();
+        bands.add(band);
```
Comment (nmc16) [inline src/main/java/net/floodlightcontroller/meters/Meters.java]: this indentation is kinda triggering me

File: src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java [hunk 3]
```diff
@@ -162,6 +171,107 @@ protected long getSpeed(NodePortTuple npt) {
...
+							
+						}else {
+						
+							//if there is no existing flow stat insert with default 0 speed
+							flowStats.put(nft, FlowBandwidth.of(sw, match, dur, bytesCount, 0,in_port)); 	
+							
+						}
+					}
+					if(debug) {
+						System.out.println("###########################################################################################################################################################################################");
```
Comment (nmc16) [inline src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java]: can you replace this system out with a logger?

File: src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java [hunk 4]
```diff
@@ -162,6 +171,107 @@ protected long getSpeed(NodePortTuple npt) {
...
+							}
+							try {
+								if(!(speed < 0)) {		
+									// if we have a valid flow we want to update the hash and then send the information to the GUI
+									
+									
+									FlowBandwidth test = FlowBandwidth.of(sw, match, dur, bytesCount, speed,in_port);
+									flowStats.put(nft,test);		
+									
+									//The SOCKET send to the gui should be implemented here
```
Comment (nmc16) [inline src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java]: Is this how we were planning on sending flow data to the GUI?

File: src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java [hunk 5]
```diff
@@ -244,7 +354,7 @@ public void init(FloodlightModuleContext context)
 
 		if (config.containsKey(INTERVAL_PORT_STATS_STR)) {
 			try {
-				portStatsInterval = Integer.parseInt(config.get(INTERVAL_PORT_STATS_STR).trim());
+				//portStatsInterval = Integer.parseInt(config.get(INTERVAL_PORT_STATS_STR).trim());
```
Comment (nmc16) [inline src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java:346]: Why'd you comment this out?

File: src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java [hunk 6]
```diff
@@ -162,6 +171,107 @@ protected long getSpeed(NodePortTuple npt) {
...
+	 * 		5. send information to the gui.
+	 * 
+	 */
+	protected class FlowStatsCollector implements Runnable {
+
+		@Override
+		public void run() {
+			Map<DatapathId, List<OFStatsReply>> flowReplies = getSwitchStatistics(switchService.getAllSwitchDpids(), OFStatsType.FLOW);
+			if(debug) {
+				System.out.print("\033[H\033[2J");// add this to clear the command line every round only works on real terminal
```
Comment (nmc16) [inline src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java]: I would probably get rid of this, especially if we plan on adding to GUI anyways

**Review by nmc16 — COMMENTED** [review]
Overall looks good, there's some whitespace stuff (like spacing and extra lines) that would be nice fixes but I'm mostly interested on how we plan on pushing stats to the GUI.

My idea was to add it to its own table, then let Luke's work deal with listening to the table and pushing it but let me know what you think. 

Also doesn't have to be added now but unit tests around the flow collection might be useful.

File: src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java (hunk 3 above)
Comment (CharlieHardwick-Kelly) [reply src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java]: will do

File: src/main/java/net/floodlightcontroller/core/web/serializers/NodeFlowTupleSerializer.java (hunk 1 above)
Comment (CharlieHardwick-Kelly) [reply src/main/java/net/floodlightcontroller/core/web/serializers/NodeFlowTupleSerializer.java]: just left there I'll delete it

File: src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java (hunk 6 above)
Comment (CharlieHardwick-Kelly) [reply src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java]: will do

File: src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java (hunk 4 above)
Comment (CharlieHardwick-Kelly) [reply src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java]: I wasn't sure, but if we want to send the data as soon as its created thats where we'll need to do it, we could also put the info into a table at this point if we wanted to do it that way

File: src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java (hunk 5 above)
Comment (CharlieHardwick-Kelly) [reply src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java:346]: It kept defaulting to the value set by the rest api

File: src/main/java/net/floodlightcontroller/meters/Meters.java (hunk 2 above)
Comment (CharlieHardwick-Kelly) [reply src/main/java/net/floodlightcontroller/meters/Meters.java]: weird it shows up as inline on my local copy

**Pull request merged**